### PR TITLE
Mk/refinements

### PR DESCRIPTION
**Pull request by kalmarek**
@fingolfin @albinahlback

### CODE REVIEW

File: src/extensions.jl [hunk 1]
```diff
@@ -0,0 +1,13 @@
+isabelian(G::Group) = false
+issolvable(G::Group) = false
+isnilpotent(G::Group) = false
+isperfect(G::Group) = false
```
Comment (albinahlback) [inline src/extensions.jl:4]: Could it better to throw an error here? It feels like we are defining that every group is non-commutative etc., don't know if I would like that behavior. But maybe...

**Comment by albinahlback** [comment]
Very nice! This one gets a :heavy_check_mark:

File: src/extensions.jl (hunk 1 above)
Comment (fingolfin) [reply src/extensions.jl:4]: Yeah, better an error - i.e. just do it like below and only provide a stub which can be documented but no actual implementation

File: README.md [hunk 2]
```diff
@@ -15,73 +15,90 @@ Due to the fact that hardly any information can be encoded in `Type`, we rely on
...
 ## `GroupElement` methods
-##### Obligatory methods
- * `Base.parent(g::GroupElement)`: return the parent object of a given group element. Parent objects of the elements of the same group must be **identical** (i.e. ===).
+#### Obligatory methods
+ * `Base.parent(g::GroupElement)`: return the parent object of a given group element. Parent objects of the elements of the same group must be **identical** (i.e. `===`).
  * `GroupsCore.parent_type(::Type{<:GroupElement})`: given the type of an element return the type of its parent.
- * `GroupsCore.istrulyequal(g::GEl, h::GEl) where GEl<:GroupElement`: return the mathematical equality of group elements; by default the standard equality `==` calls this function.
- * `GroupsCore.hasorder(g::GroupElement)`: return `true` if `g` has finite order (without computing it)`.
+ * `GroupsCore.==(g::GEl, h::GEl) where GEl<:GroupElement`: return the mathematical equality of group elements;
+ * `GroupsCore.isfiniteorder(g::GroupElement)`: return `true` if `g` has finite order (possibly without computing it)`. If `isfiniteorder(g)` returns `false`, `order(g)` is required to throw `GroupsCore.InfiniteOrder` exception.
```
Comment (fingolfin) [inline README.md:43]: Is that what we settled on? I mean, surely `order(BigInt, g)` or `order(fmpz, g)` etc. should do that; but wasn't the option of using something like https://gist.github.com/kalmarek/640630a53d5837a5c7fe401eb2f047bd for `order` also still on the table, and/or returning something like a `Union{BigInt, SomeTypeRepresentingInfiniteOrder}` still an option?

Comment (fingolfin) [reply README.md:43]: SomeTypeRepresentingInfiniteOrder could be a bunch of things: a Float type and we use `inf`; our own type representing infinity; we could also (ab)use `Nothing`  ("this group/element has no finite order, so we can return nothing")

File: README.md [hunk 3]
```diff
@@ -15,73 +15,90 @@ Due to the fact that hardly any information can be encoded in `Type`, we rely on
...
+ * `Base.one(g::GroupElement)` → `one(parent(g))`
+ * `GroupsCore.order(g)`, `order(I::Type{<:Integer}, g)` (naive implementation)
  * `Base.literal_pow(::typeof(^), g, Val{-1})` → `inv(g)`
- * `Base.:(/)(g, h)` → `g*inv(h)`
- * `Base.conj(g, h)`, `Base.:(^)(g, h)` → `inv(h)*g*h`
- * `Base.comm(g, h)` → `inv(h)*inv(g)*h*g` and its `Vararg` (`foldl`) version.
- * `Base.:(==)(g,h)` → `GroupsCore.istrulyequal(g, h)`
+ * `Base.:(/)(g, h)` → `g*h^-1`
+ * `Base.conj(g, h)`, `Base.:(^)(g, h)` → `h^-1*g*h`
+ * `Base.comm(g, h)` → `h^-1*g^-1*h*g` and its `Vararg` (`foldl`) version.
```
Comment (fingolfin) [inline README.md:57]: The convention used in GAP is:
```suggestion
 * `Base.comm(g, h)` → `g^-1*h^-1*g*h` and its `Vararg` (`foldl`) version.
```
This is also what Wikipedia lists (not that this means too much) and also what I see in books on group theory most frequently; it is consistent with right actions.

People preferring left actions typically also prefer defining `[g,h] = g*h*g^{-1}*h^{-1}`.

I am not aware of any source using "your" definition, but of course they still may exist :-)

Comment (fingolfin) [reply README.md:57]: OK, `g^-1*h^-1*g*h` is also what you write for `comm!` so perhaps we can agree on that ;-).

File: README.md (hunk 2 above)
Comment (kalmarek) [reply README.md:43]: We left it to better times: see the discussion in #8

File: README.md (hunk 3 above)
Comment (kalmarek) [reply README.md]: yes, that was bug in the readme

File: README.md [hunk 4]
```diff
@@ -15,73 +15,90 @@ Due to the fact that hardly any information can be encoded in `Type`, we rely on
...
-   which is a "best effort", cheap computation of length of the group iterator. For practical reasons the largest group you could possibly iterate over is of order ~`factorial(19)` (which still fits into `Int`).
+Additionally the following assumptions are placed on the iteration:
+ * `first(G)` must be the identity
+ * iteration over an infinite group should exhaust every fixed radius ball (in word-length metric) around the identity in finite time.
 
-##### Obligatory methods
+#### Obligatory methods
  * `Base.one(G::Group)`: return the identity of the group
- * `GroupsCore.order(::Type{I}, G::Group) where I<:Integer`: the order of `G` returned as an instance of `I`; only arbitrary size integers are required to return mathematically correct answer.
+ * `GroupsCore.order(::Type{I}, G::Group) where I<:Integer`: the order of `G` returned as an instance of `I`; only arbitrary size integers are required to return mathematically correct answer. An infinite group must throw `GroupsCore.InfiniteOrder` exception.
```
[the remark this answers is not in the document]
Comment (kalmarek) [reply README.md:34]: It is defined based on the iterator type, with the default to `false`. We could possibly throw an error if that one returns `Base.SizeUnknown()`, to stay on the absolute correctness side. Although we write specifically that by default groups are infinite, unless you change the iterator type to explicitly request something finite.

**Comment by kalmarek** [comment]
@fingolfin I'd be happiest to merge this as is and release `v0.1.0`;

We can have further discussions and refinements for subsequent versions.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kalmarek) [reply README.md:34]: ok, afer brief moment I went for throwing an error asking user kindly to implement his own `Base.isfinite` ;)

**Pull request merged**